### ISSUE TITLE
Batch correction - strange convergence

### ISSUE THREAD

**Issue by hanbin973** [willtownes/glmpca#30]
Hi, I found out something strange when using batch correction.
These are the two result from Muraro (pancreas) and MacParland (liver) datasets.
This is from Muraro
![image](https://user-images.githubusercontent.com/17215340/115637727-cfda5880-a34b-11eb-8265-499f7d697e42.png)
and this is from Macparland
![image](https://user-images.githubusercontent.com/17215340/115637762-ee405400-a34b-11eb-89f1-6ae4a434e474.png)
Each had 4 and 5 donors respectively.
The commnad I used is the following:

`RunGLMPCA(obj, X=model.matrix(as.formula('~factor(batch)'))`

The result is less strange when I remove the batch covariate matrix X from the model but the result is without batch correction.
For example, the Muraro et al.'s pancreas dataset
![image](https://user-images.githubusercontent.com/17215340/115638941-5d1eac80-a34e-11eb-8f57-d3d6e391625d.png)
the batch effect is present (in orange) but the overall result look less weird.

Any solutions possible at this time?

**Comment by willtownes** [comment]
Hello and thanks for trying glmpca. That is indeed a weird result! To be honest I have not thoroughly tested the batch correction feature. However, I suspect there is a numerical problem. By default, glmpca will subtract off the column means of the covariates matrix X and tack on a column of all ones. Since your provided model.matrix is going to already have a column of ones, I believe this is creating a duplicate column of ones. This collinearity is likely to cause numerical problems. Please try the following and let me know what happens:
```R
X<-model.matrix(as.formula('~factor(batch)')
RunGLMPCA(obj, X=X[,2:ncol(X)])
```
In case you are interested here is the [relevant part of the code](https://github.com/willtownes/glmpca/blob/master/R/initialization.R#L197).

**Comment by willtownes** [comment]
Note to self, it would be good to check for this condition and drop the extra column automatically in the initialization code.

**Comment by hanbin973** [comment]
I have tested after dropping the first column as you suggested but getting something nearly the same:

![image](https://user-images.githubusercontent.com/17215340/115651052-68c99d80-a365-11eb-9841-9485dc17f9b9.png)

The code:
```
X <- as.matrix(model.matrix(as.formula('~factor(donor)'), pbmc@meta.data))
X <- X[,2:ncol(X)]
pbmc_gpca <- RunGLMPCA(pbmc, L=30, 
                       X=X, 
                       sz=size_factor)
```

**Comment by willtownes** [comment]
Yikes, still pretty weird! What happens if you set L=2 and/or set the learning rate to something small like .0001 ? To do the latter, you can pass it in as the "ctl" list `ctl=list(lr=.001)`. The learning rate default is normally 0.1. And how many features (genes) are in the pbmc matrix?

**Comment by hanbin973** [comment]
I supplied 2000 features (feature selection was performed using Seurat's default vst) . By the way, I really thank you for the super fast replies. Trying adjusting the lr.

p.s. The resulting plots are GLM-PCA supplied to t-SNE (python openTSNE) with default settings.

**Comment by willtownes** [comment]
Ohhh, that makes more sense that they are from tSNE. Please try modifying the perplexity parameter on the tSNE to see if that makes a difference. It might also be worth trying the Rtsne package and one of the UMAP implementations with glmpca initialization just in case there is a bug in openTSNE. I have seen tSNE sometimes do weird stuff even with standard PCA initializations. If you run GLM-PCA with L=2 you can directly visualize the latent space without needing to pass to tSNE, so that should also be helpful to compare.

**Comment by hanbin973** [comment]
![image](https://user-images.githubusercontent.com/17215340/115655667-b696d380-a36e-11eb-8e9c-9f9b918a389e.png)
Adjusting the learning rate didn't help (X with dropped intercepts, openTSNE)

The next figure was produced after reducing the perplexity but still the same
![image](https://user-images.githubusercontent.com/17215340/115655834-083f5e00-a36f-11eb-9492-f725ca69b08b.png)

This one is a pure GLM-PCA with L=2
![image](https://user-images.githubusercontent.com/17215340/115656311-01fdb180-a370-11eb-9c7c-7ea55fcfce87.png)
which looks OK.

This one is GLM-PCA (X with dropped interceps, reduced lr to 0.001, Seurat TSNE) + TSNE which is clearly strange
![image](https://user-images.githubusercontent.com/17215340/115656458-38d3c780-a370-11eb-8e0e-98d982b0d00e.png)

This one is GLM-PCA + UMAP (other parameters are the same with the previous TSNE setting)
![image](https://user-images.githubusercontent.com/17215340/115656564-615bc180-a370-11eb-82dc-580e7520ac50.png)

I think learning rate nor the specific tSNE implementation nor the UMAP implementation is the problem.
Probably L>5~10 + batch matrix is not working properly in the current implementation.

**Comment by chaudhry123** [comment]
Same problem in UMAP. Any updates?

**Comment by willtownes** [comment]
Sorry, I haven't had a chance to work on this, and probably won't for some time. If anyone else gets a chance to figure it out I'd be curious to know what's going on!

**Comment by hanbin973** [comment]
Not sure but may be some initialization scheme through (computationally much cheaper) PCA might help.
Is there an option in GLM-PCA that takes initial latent coordinates?

**Comment by willtownes** [comment]
Yes, there is an option called `init` where you can pass in pre-estimated factors, loadings, or both. I have tried it a few times but usually it takes just as long (if not longer) to converge than using a random initialization.
